File: test/poller.test.ts
import { describe, expect, test } from "@effect/vitest";
import { Effect, Layer } from "effect";

import { RuleActionExecutor, compileRules } from "../src/dsl/compiler.ts";
import type { QueryEnvironment } from "../src/dsl/environment.ts";
import { GitHubClient, type GitHubNotification } from "../src/github.ts";
import { NotificationPoller, NotificationPollerLive } from "../src/poller.ts";

const notification: GitHubNotification = {
  id: "1",
  unread: true,
  reason: "mention",
  updated_at: "2026-09-19T10:00:00Z",
  last_read_at: null,
  repository: { full_name: "luxass/cloudflare-workers" },
  subject: { title: "Review production polling", type: "PullRequest", url: null },
};

describe("NotificationPoller", () => {
  test("polls GitHub and executes matching actions through services", async () => {
    const actions: string[] = [];
    const rules = await Effect.runPromise(
      compileRules([
        {
          name: "mentions",
          when: 'notification.reason == "mention"',
          actions: [{ type: "read" }],
        },
      ]),
    );
    const githubLayer = Layer.succeed(GitHubClient, {
      listNotifications: () =>
        Effect.succeed({
          kind: "updated" as const,
          notifications: [notification],
          pollAfterMs: 65_000,
          truncated: false,
        }),
      markThreadRead: () => Effect.void,
      markThreadDone: () => Effect.void,
      deleteThreadSubscription: () => Effect.void,
      getSubject: () => Effect.succeed(null),
    });
    const executor: RuleActionExecutor["Service"] = {
      execute: (action, _environment: QueryEnvironment) => {
        actions.push(action.type);
        return Effect.void;
      },
    };
    const layer = NotificationPollerLive(rules).pipe(
      Layer.provide(Layer.succeed(RuleActionExecutor, executor)),
      Layer.provide(githubLayer),
    );

    await Effect.runPromise(
      Effect.gen(function* () {
        const poller = yield* NotificationPoller;
        yield* poller.poll("startup");
      }).pipe(Effect.provide(layer)),
    );

    expect(actions).toEqual(["read"]);
  });

  test("fetches a subject only when notification fields leave a rule undecided", async () => {
    const subjectUrl = "https://api.github.com/repos/a/b/pulls/1";
    const notifications = [
      {
        ...notification,
        id: "irrelevant",
        reason: "mention",
        subject: { ...notification.subject, url: subjectUrl },
      },
      {
        ...notification,
        id: "candidate",
        reason: "review_requested",
        subject: { ...notification.subject, url: subjectUrl },
      },
    ];
    const fetchedSubjects: string[] = [];
    const actions: string[] = [];
    const rules = await Effect.runPromise(
      compileRules([
        {
          name: "stale-review",
          when: 'notification.reason == "review_requested" and subject.reviewPending == false',
          actions: [{ type: "done" }],
        },
      ]),
    );
    const layer = NotificationPollerLive(rules).pipe(
      Layer.provide(
        Layer.succeed(RuleActionExecutor, {
          execute: (action) => {
            actions.push(action.type);
            return Effect.void;
          },
        }),
      ),
      Layer.provide(
        Layer.succeed(GitHubClient, {
          listNotifications: () =>
            Effect.succeed({
              kind: "updated" as const,
              notifications,
              pollAfterMs: 0,
              truncated: false,
            }),
          markThreadRead: () => Effect.void,
          markThreadDone: () => Effect.void,
          deleteThreadSubscription: () => Effect.void,
          getSubject: (url) =>
            Effect.sync(() => {
              fetchedSubjects.push(url);
              return { state: "open", merged: false, author: "octocat", reviewPending: false };
            }),
        }),
      ),
    );

    await Effect.runPromise(
      Effect.gen(function* () {
        const poller = yield* NotificationPoller;
        yield* poller.poll("startup");
      }).pipe(Effect.provide(layer)),
    );

    expect(fetchedSubjects).toEqual([subjectUrl]);
    expect(actions).toEqual(["done"]);
  });

  test("reuses Last-Modified for conditional notification polls", async () => {
    const requests: Array<{ readonly lastModified?: string }> = [];
    let pollCount = 0;
    const rules = await Effect.runPromise(compileRules([]));
    const layer = NotificationPollerLive(rules).pipe(
      Layer.provide(Layer.succeed(RuleActionExecutor, { execute: () => Effect.void })),
      Layer.provide(
        Layer.succeed(GitHubClient, {
          listNotifications: (options) => {
            requests.push(options ?? {});
            pollCount += 1;
            return Effect.succeed(
              pollCount === 1
                ? {
                    kind: "updated" as const,
                    notifications: [],
                    pollAfterMs: 0,
                    lastModified: "Sat, 19 Sep 2026 10:00:00 GMT",
                    truncated: false,
                  }
                : { kind: "not-modified" as const, pollAfterMs: 0 },
            );
          },
          markThreadRead: () => Effect.void,
          markThreadDone: () => Effect.void,
          deleteThreadSubscription: () => Effect.void,
          getSubject: () => Effect.succeed(null),
        }),
      ),
    );

    await Effect.runPromise(
      Effect.gen(function* () {
        const poller = yield* NotificationPoller;
        yield* poller.poll("startup");
        yield* poller.poll("schedule");
      }).pipe(Effect.provide(layer)),
    );

    expect(requests).toEqual([{}, { lastModified: "Sat, 19 Sep 2026 10:00:00 GMT" }]);
  });

  test("does not execute actions for an unchanged poll", async () => {
    const actions: string[] = [];
    const rules = await Effect.runPromise(compileRules([]));
    const layer = NotificationPollerLive(rules).pipe(
      Layer.provide(
        Layer.succeed(RuleActionExecutor, {
          execute: (action) => {
            actions.push(action.type);
            return Effect.void;
          },
        }),
      ),
      Layer.provide(
        Layer.succeed(GitHubClient, {
          listNotifications: () =>
            Effect.succeed({ kind: "not-modified" as const, pollAfterMs: 65_000 }),
          markThreadRead: () => Effect.void,
          markThreadDone: () => Effect.void,
          deleteThreadSubscription: () => Effect.void,
          getSubject: () => Effect.succeed(null),
        }),
      ),
    );

    await Effect.runPromise(
      Effect.gen(function* () {
        const poller = yield* NotificationPoller;
        yield* poller.poll("schedule");
      }).pipe(Effect.provide(layer)),
    );

    expect(actions).toEqual([]);
  });
});

File: internal/app/poller_test.go
package app

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/luxass/github-notifications/internal/github"
)

func TestLazySubjectFetching(t *testing.T) {
	const subjectPath = "/repos/a/b/pulls/1"
	var subjectFetches, doneMutations int
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		switch request.URL.Path {
		case subjectPath:
			subjectFetches++
			if _, err := fmt.Fprint(writer, `{"state":"open","user":{"login":"octocat"},"requested_reviewers":[]}`); err != nil {
				t.Error(err)
			}
		case "/notifications/threads/candidate":
			doneMutations++
			writer.WriteHeader(http.StatusNoContent)
		default:
			http.NotFound(writer, request)
		}
	}))
	defer server.Close()

	config := Config{
		Server: ServerConfig{Host: "127.0.0.1", Port: 3001, Schedule: "0 * * * * *"},
		GitHub: GitHubConfig{TokenEnv: "GITHUB_TOKEN", MaxPages: 5},
		Rules: []RuleConfig{{
			Name:    "stale-review",
			When:    `notification.reason == "review_requested" and subject.reviewPending == false`,
			Actions: []ActionConfig{{Type: "done"}},
		}},
	}
	client := github.NewClientWithBaseURL("token", server.Client(), server.URL)
	poller, err := NewPoller(config, client, slog.New(slog.NewTextHandler(io.Discard, nil)))
	if err != nil {
		t.Fatal(err)
	}

	url := server.URL + subjectPath
	irrelevant := testNotification("irrelevant", "mention", url)
	if _, err := poller.processNotification(context.Background(), irrelevant); err != nil {
		t.Fatal(err)
	}
	if subjectFetches != 0 {
		t.Fatalf("unrelated notification fetched subject %d times", subjectFetches)
	}

	candidate := testNotification("candidate", "review_requested", url)
	matched, err := poller.processNotification(context.Background(), candidate)
	if err != nil {
		t.Fatal(err)
	}
	if matched != 1 || subjectFetches != 1 || doneMutations != 1 {
		t.Fatalf("unexpected lazy evaluation: matched=%d fetches=%d mutations=%d", matched, subjectFetches, doneMutations)
	}
}

func testNotification(id, reason, subjectURL string) github.Notification {
	var notification github.Notification
	notification.ID = id
	notification.Unread = true
	notification.Reason = reason
	notification.UpdatedAt = time.Date(2026, 9, 19, 10, 0, 0, 0, time.UTC)
	notification.Repository.FullName = "a/b"
	notification.Subject.Title = "review"
	notification.Subject.Type = "PullRequest"
	notification.Subject.URL = &subjectURL
	return notification
}

func TestPollDoesNotAdvanceLastModifiedAfterActionFailure(t *testing.T) {
	const lastModified = "Sat, 19 Sep 2026 10:00:00 GMT"
	var actionAttempts int
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		switch request.URL.Path {
		case "/notifications":
			if request.Header.Get("If-Modified-Since") != "" {
				t.Errorf("unexpected conditional request after failed poll: %q", request.Header.Get("If-Modified-Since"))
			}
			writer.Header().Set("Last-Modified", lastModified)
			writer.Header().Set("Content-Type", "application/json")
			_, _ = fmt.Fprint(writer, `[{"id":"thread-1","unread":true,"reason":"mention","updated_at":"2026-09-19T10:00:00Z","repository":{"full_name":"a/b","private":false},"subject":{"title":"hello","type":"Issue","url":null}}]`)
		case "/notifications/threads/thread-1":
			actionAttempts++
			if actionAttempts == 1 {
				writer.WriteHeader(http.StatusInternalServerError)
				return
			}
			writer.WriteHeader(http.StatusNoContent)
		default:
			http.NotFound(writer, request)
		}
	}))
	defer server.Close()

	config := Config{
		Server: ServerConfig{Host: "127.0.0.1", Port: 3001, Schedule: "0 * * * * *"},
		GitHub: GitHubConfig{TokenEnv: "GITHUB_TOKEN", MaxPages: 5},
		Rules: []RuleConfig{{
			Name:    "mark-mention-done",
			When:    `notification.reason == "mention"`,
			Actions: []ActionConfig{{Type: "done"}},
		}},
	}
	poller, err := NewPoller(config, github.NewClientWithBaseURL("token", server.Client(), server.URL), slog.New(slog.NewTextHandler(io.Discard, nil)))
	if err != nil {
		t.Fatal(err)
	}

	if _, err := poller.Poll(context.Background(), "test"); err == nil {
		t.Fatal("expected first action to fail")
	}
	if poller.lastModified != "" {
		t.Fatalf("lastModified advanced after failed poll: %q", poller.lastModified)
	}
	if _, err := poller.Poll(context.Background(), "test"); err != nil {
		t.Fatalf("second poll failed: %v", err)
	}
	if poller.lastModified != lastModified || actionAttempts != 2 {
		t.Fatalf("notification was not retried: lastModified=%q actionAttempts=%d", poller.lastModified, actionAttempts)
	}
}

func TestFailedSubjectFetchDoesNotRunActions(t *testing.T) {
	const subjectPath = "/repos/a/b/pulls/1"
	var subjectFetches, actionAttempts int
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		switch request.URL.Path {
		case subjectPath:
			subjectFetches++
			writer.WriteHeader(http.StatusInternalServerError)
		case "/notifications/threads/candidate":
			actionAttempts++
			writer.WriteHeader(http.StatusNoContent)
		default:
			http.NotFound(writer, request)
		}
	}))
	defer server.Close()

	config := Config{
		Server: ServerConfig{Host: "127.0.0.1", Port: 3001, Schedule: "0 * * * * *"},
		GitHub: GitHubConfig{TokenEnv: "GITHUB_TOKEN", MaxPages: 5},
		Rules: []RuleConfig{{
			Name:    "not-pending-review",
			When:    `notification.reason == "review_requested" and subject.reviewPending == false`,
			Actions: []ActionConfig{{Type: "done"}},
		}},
	}
	poller, err := NewPoller(config, github.NewClientWithBaseURL("token", server.Client(), server.URL), slog.New(slog.NewTextHandler(io.Discard, nil)))
	if err != nil {
		t.Fatal(err)
	}

	matched, err := poller.processNotification(context.Background(), testNotification("candidate", "review_requested", server.URL+subjectPath))
	if err != nil {
		t.Fatal(err)
	}
	if matched != 0 || subjectFetches != 1 || actionAttempts != 0 {
		t.Fatalf("failed subject fetch ran action: matched=%d fetches=%d actions=%d", matched, subjectFetches, actionAttempts)
	}
}
